### DISCUSSION TITLE
Release binaries in weird format

### DISCUSSION THREAD

**Discussion by lordofscripts** [community/community#137202] [Actions] [answered]
I am new to actions. My [WipeChromium](https://GitHub.com/lordofscripts/wipechromium/) project has a `crossbuild.yml` **action** in it that builds my GO project the usual way for 3 platforms and the published the resulting binaries to a Draft Release Note.

There are 2 problems with it:

- Even though it is one workflow, on every trigger 2 release drafts get produced. The first with 2 binaries and the 2nd with the remaining binary. Why does it splits it?  At least for now I can live with it but an explanation would be nice.
- My major problem is that the resulting binaries are not executables, but they appear not to be neither zip nor tar nor gzip. When I use the `file` command on it, it says the binaries are **AR** archives! Why? The output of go `build -o` file is a binary, not a linkable object file! The Member in the archive doesn't even have the name I gave in the `-o` option! The Member in the archive shows as `_go_.o` so I am puzzled.

**Comment by spenserblack** [comment, marked as answer]
Because the workflow is generating 2 releases, both with incomplete data, I'm curious if there's a race condition that's caused by running the builds in parallel using a matrix. Because it's super easy to cross-compile with Go, I'd recommend trying out sequential builds instead of parallel builds. That might honestly be preferable since you should end up using less action minutes that way (you'd only build 3 times instead of doing the whole job 3 times). I believe you could also get around this issue by not using an action that *creates* releases in a matrix, and ensuring that you only create a release once. This could be done by having 1 job to create the draft release, then a matrix job to build and upload release assets. But, to keep it simple for now, try removing the matrix and instead writing the build step 3 times with different values in `env:`. But, whichever way you choose, the goal is to run the `action-gh-release` action only once.

As far as the filetype issue goes, I believe it's because your `package main` is *not* at root level, and you're running `go build`, not `go build ./...`. When you run `go build` at the root level, it's building the root-level *library* files, not the binary files that are nested. The same thing should happen if you run `go build -o not-a-binary` locally. A common command is `go build ./...`, where `...` is special and means "all files here and below", basically. In other words, `go build ./...` should build your entire module, not just the package at the current directory.

I'd recommend ensuring that you can build your project correctly locally on your own device, then roughly copying those commands to your workflow.

**Reply by lordofscripts** [reply]
Thanks for the tips and explanations. 

The `go build ./... -o name` gives a weird error. However, I fixed it by being explicit: `go build -o NAME cmd/wiper/*.go`. As you pointed out, that should build the actual application executable rather than a module's library object file 👍.

Since I don't know how to not use a matrix, the workflow (working on it) will still use a matrix but I altered the strategy to _limit the parallel runs_ to just one: `max-parallel: 1` underneath `strategy:`.

Will be trying it out in a few minutes.

**Reply by lordofscripts** [reply]
@spenserblack Thanks for your exceptional guidance and willingness to help others walk in unfamiliar territory.

I got the workflow to do the cross-build properly, sequentially but it works:

https://github.com/lordofscripts/wipechromium/blob/main/.github%2Fworkflows%2Fcrossbuild.yml

Perhaps you have some ideas to point me in the right direction to make the following workflow improvements:

- [ ] Windows binary doesn't have `.exe` extension.
- [ ] None of the attached assets (cross-compilation binaries) were uploaded in compressed form (`.zip` for Windows, else `.tar.gz`)

**Reply by spenserblack** [reply]
> Since I don't know how to not use a matrix

It's going to be easier than you're expecting, probably :slightly_smiling_face: Just write the step 3 times with different variables. While this can be pretty unmaintainable if you have a lot of target OSs/architectures, with 3 it wouldn't be so bad.

```yaml
- run: go build "myexe-$GOOS-GOARCH" path/to/myexe.go
  env:
    GOOS: linux
    GOARCH: 386
# ...
- run: go build "myexe-$GOOS-GOARCH.exe" path/to/myexe.go
  env:
    GOOS: linux
    GOARCH: 386
```

If this starts to get too unmaintainable, you can still keep it as only 1 running job by using a script with a loop to build the assets (see [gh-extension-precompile](https://github.com/cli/gh-extension-precompile)'s script for an example).

For adding the `.exe` extension, I just do a simple check if the job is running on Windows. If so, append `.exe` to the output filename. You can check out [the GitHub CLI's extension precompile action](https://github.com/cli/gh-extension-precompile) for an example of the logic used to compile a Go executable.

If you want to compress the assets, you'll have to do that yourself by including a call to `tar` or `zip` in your workflow. For example:

```yaml
- run: zip -r myexe.zip path/to/myexe.exe
- run: tar czf myexe.tar.gz path/to/myexe
```
